core/dbt/adapters/base/impl.py

```diff
@@ -1125,6 +1125,37 @@ def get_rows_different_sql(
 
         return sql
 
+    def validate_incremental_strategy(self):
+        return ["append", "delete+insert"]
+
+    @available.parse_none
+    def get_incremental_strategy_macro(self, strategy: str):
+        # Construct macro_name from strategy name
+        if strategy is None:
+            strategy = "default"
+        strategy = strategy.replace("+", "_")
+        macro_name = f"get_incremental_{strategy}_sql"
+        # get a macro and return it
+        manifest = self._macro_manifest
+        macro = manifest.find_macro_by_name(macro_name, self.config.project_name, None)
+        if not macro:
+            raise RuntimeException(
+                'dbt could not find a macro with the name "{}" in {}'.format(
+                    macro_name, self.config.project_name
+                )
+            )
+
+        from dbt.context.providers import generate_runtime_macro_context
+
+        macro_context = generate_runtime_macro_context(
+            macro=macro,
+            config=self.config,
+            manifest=manifest,  # type: ignore[arg-type]
+            package_name=self.config.project_name,
+        )
+        macro_function = MacroGenerator(macro, macro_context)
+        return macro_function
```

Comment:
This code feels rough for us, but on the bright side, we're the only ones that have to deal with it. It's nicely cordoned off from user space. Up to us to decide whether this approach feels cleaner than the macro discovery/lookup/"dispatch" that happens for snapshots today ([here](https://github.com/dbt-labs/dbt-core/blob/5b1bc72ae11a36031b88d4d26a4ead3be87b3e50/core/dbt/include/global_project/macros/materializations/snapshots/strategies.sql#L1-L32)).